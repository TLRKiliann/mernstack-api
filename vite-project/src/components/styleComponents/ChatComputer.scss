/*
purple
orange
*/

.div--imgcompute {
  position: fixed;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
  opacity: 0.7;
  background: linear-gradient(180deg, purple, orange);
  z-index: -1;

  .img--chatcomputerroom {
    opacity: 0.1;
  }
}

.title--computerroom {
  .compter--roomdb {
    margin-top: 120px;
    margin-left: 20px;
    font-family: "Acme", sans-serif;
    font-size: 2.4rem;
    color: whitesmoke;
    text-shadow: 2px 2px 1px purple,
    4px 3px 2px #242424;
  }
}

.computer--room {
  margin-top: 20px;
  
  .links--computer {
    font-family: "DejaVu Sans", sans-serif;
    cursor: pointer;

    .span--linktoroom {
      padding: 20px 0px;
      padding-left: 20px;
      margin: auto;
      margin-right: 20px;
      margin-left: 20px;
      display: flex;

      background-color: rgba(85, 45, 255, 0.6);
      border: 2mm outset rgba(85, 45, 255, 0.2);
      box-shadow: 0px 0px 4px slategrey;
      border-radius: 10px;

      text-decoration: none;
      color: whitesmoke;
      &:hover {
        background-color: rgba(85, 45, 255, 0.4);
        color: orange;
      }
      &:active {
        background-color: purple;
        color: lightgreen;
      }
    }
  }
}