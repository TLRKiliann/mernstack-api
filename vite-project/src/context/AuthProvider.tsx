import React, { createContext, ReactNode,
  useState, useContext } from 'react'

type Props = {
  children: ReactNode
}

type AuthContextProps = {
  auth: object
  setAuth: React.Dispatch<React.SetStateAction<object>>
  username: string
  setUserName: React.Dispatch<React.SetStateAction<string>>
  switchLogin: boolean
  setSwitchLogin: React.Dispatch<React.SetStateAction<boolean>>
  toggle: () => void
  eraseAll: () => void
}

const AuthContext = createContext({} as AuthContextProps);

export const useAuthLogin: React.FC = () => {
  return useContext(AuthContext)
}

export const AuthProvider: React.FC = ({ children }: Props) => {

  const [auth, setAuth] = useState<object>({});
  const [switchLogin, setSwitchLogin] = useState(true);

  const [username, setUserName] = useState("");

  const toggle = () => {
      setSwitchLogin(!switchLogin);
  };

  const eraseAll = () => {
      setUserName("");
      setAuth("");
  };

  return (
    <AuthContext.Provider value={{
      auth,
      setAuth,
      username,
      setUserName,
      switchLogin,
      setSwitchLogin,
      toggle,
      eraseAll
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};