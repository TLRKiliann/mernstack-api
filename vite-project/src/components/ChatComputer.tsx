import React, { useState, useEffect } from 'react'
import { useParams, useNavigate } from 'react-router-dom'
import { useAuthLogin } from '../context/AuthProvider'
import usePersonnalHook from '../hook/personnal.hook'
import serviceRouting from '../services/serviceRouting'
import { db_computers } from '../models/db_computers'
import { db_computeOne } from '../models/db_computeList'
import { db_computeTwo } from '../models/db_computeList'
import { db_computeThree } from '../models/db_computeList'
import { db_computeFour } from '../models/db_computeList'
import { db_computeFive } from '../models/db_computeList'
import { db_computeSix } from '../models/db_computeList'
import { db_computeSeven } from '../models/db_computeList'
import { db_computeEight } from '../models/db_computeList'
import { db_computeNine } from '../models/db_computeList'
import { db_computeTen } from '../models/db_computeList'
import { db_computeEleven } from '../models/db_computeList'
import { db_computeTwelve } from '../models/db_computeList'
import { computerType } from '../models/computerType'
import img_1 from '../assets/hdd.jpg'
import img_2 from '../assets/cpu.jpg'
import img_3 from '../assets/ram.jpg'
import img_4 from '../assets/motherboard.png'
import img_5 from '../assets/systemd.jpg'
import img_6 from '../assets/process.jpg'
import img_7 from '../assets/world_bin.png'
import img_8 from '../assets/reqres.jpg'
import img_9 from '../assets/firewall.jpg'
import img_12 from '../assets/Cybersecurity.jpg'
import './styleComponents/ChatComputer.scss'


const ChatComputer: React.FC = () => {

  const {id} = useParams<{id?: string}>();
  const { username } = useAuthLogin()
  const Navigate = useNavigate()
  
  const users = usePersonnalHook()
  
  //console.log(users, "- object users")

  const [userRoom, setUserRoom] = useState<Array<UserType>>([])
  //console.log(userRoom, 'userRoom state')
  const [computerDb, setComputerDb] = useState<string>("")
  const [imgBg, setImgBg] = useState<string>("")
  const [links, setLinks] = useState<Array<computerType>>([])

  const handleNavigation = (link: string) => {
    const myTimeOut = setTimeout(() => {
      Navigate(`/computerroom/${link}`)
    }, 1000)
  }

  const handleSetUserRoom = (link: string) => {
    const user = users?.find(user => user.firstName === username)
    const id: number = user.id;
    const changeConfRoom = {...user, isConnected: true}
    const addRoomUser = {
      id: user.id,
      img: user.img,
      firstName: username,
      lastName: user.lastName,
      age: user.age,
      email: user.email,
      location: user.location,
      gender: user.gender,
      mainroom: computerDb,
      room: link,
      isConnected: true,
      signalRecieve: user.signalRecieve,
      messagebox: user.messagebox,
      returnConfirm: user.returnConfirm
    }
    //console.log(addRoomUser.id, 'by id')
    if (user) {
      serviceRouting
        .updateRoomName(id, changeConfRoom)
        .then(initialUserRoom => {
          setUserRoom(users?.map(user => user.id === id ? {
            id: user.id,
            img: user.img,
            firstName: username,
            lastName: user.lastName,
            age: user.age,
            email: user.email,
            location: user.location,
            gender: user.gender,
            mainroom: computerDb,
            room: link,
            isConnected: true,
            signalRecieve: user.signalRecieve,
            messagebox: user.messagebox,
            returnConfirm: user.returnConfirm
            } : user
          ))
        })
        .catch((error) => {
          setUserRoom(users?.map(user => user.firstName !== username))
          alert(`Register name Room issue: ${user.firstName} not found !`)
        })
      handleNavigation(link)
    } else {
      alert("ERROR !!!")
    }
  }

  useEffect(() => {
    switch(id?.toString()) {
      case "1":
        setComputerDb("Hard Disk - SSD")
        setImgBg(img_1)
        setLinks(db_computeOne)
        break
      case "2":
        setComputerDb("CPU")
        setImgBg(img_2)
        setLinks(db_computeTwo)
        break
      case "3":
        setComputerDb("RAM")
        setImgBg(img_3)
        setLinks(db_computeThree)
        break
      case "4":
        setComputerDb("Drivers")
        setImgBg(img_4)
        setLinks(db_computeFour)
        break
      case "5":
        setComputerDb("Systemd")
        setImgBg(img_5)
        setLinks(db_computeFive)
        break
      case "6":
        setComputerDb("Process")
        setImgBg(img_6)
        setLinks(db_computeSix)
        break
      case "7":
        setComputerDb("Client-Server")
        setImgBg(img_7)
        setLinks(db_computeSeven)
        break
      case "8":
        setComputerDb("Protocols")
        setImgBg(img_7)
        setLinks(db_computeEight)
        break
      case "9":
        setComputerDb("Request - Response")
        setImgBg(img_8)
        setLinks(db_computeNine)
        break
      case "10":
        setComputerDb("Fire-Wall")
        setImgBg(img_9)
        setLinks(db_computeTen)
        break
      case "11":
        setComputerDb("Proxy - VPN")
        setImgBg(img_12)
        setLinks(db_computeEleven)
        break
      case "12":
        setComputerDb("Cyber-Security")
        setImgBg(img_12)
        setLinks(db_computeTwelve)
        break
      default:
        console.log("end of switch loop !")
        break
    }
  }, [])

  return(
    <div>
      
      <div data-testid="chatctestid" className="div--imgcompute">
        <img
          src={imgBg}
          width="100%"
          height="100%"
          className="img--chatcomputerroom"
          alt={imgBg}
        />
      </div>
      
      <div className="title--computerroom">
        <h1 className="compter--roomdb">{computerDb}</h1>
      </div>

      <div className='compter--room'>
        {Object.values(links).map((val) => (
          <h2 key={val.title} className="links--computer">

            <span
              onClick={() => handleSetUserRoom(val.link)} 
              className="span--linktoroom"
            >
              {val.link}
            </span>

          </h2>
          ))
        }
      </div>

    </div>
  )
}

export default ChatComputer;
