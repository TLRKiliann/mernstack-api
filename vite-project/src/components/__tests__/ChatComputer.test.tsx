import React from "react"
import {MemoryRouter} from 'react-router-dom'
import { useAuthLogin } from '../../context/AuthProvider'
import { screen, render, getByTestId } from '@testing-library/react'
import { test, expect, vi, beforeEach, describe } from 'vitest'
import '@testing-library/jest-dom'
import { create } from 'react-test-renderer'
import ChatComputer from "../ChatComputer"
import handleSetUserRoom from "../ChatComputer"
import handleNavigation from "../ChatComputer"

test("ChatComputer context testing", () => {
  beforeEach(async (context) => {
    context.foo = 'bar'
  })
  test('Profile state', ({ foo }) => {
    console.log(foo) // 'bar'
  })
})

test('MatchSnapShot test from ChatComputer', () => {
  const treeChatComp = create(
    <MemoryRouter>
      <ChatComputer />
    </MemoryRouter>
  )
  expect(treeChatComp.toJSON()).toMatchSnapshot()
})

test('ChatComputer handleSetUserRoom(fn) return link', () => {
  const beverage = {link: "newRoom" }
  const handleSetUserRoom = vi.fn(beverage => {beverage})
  handleSetUserRoom(beverage)
  expect(handleSetUserRoom).toHaveReturned("newRoom")
  expect(handleSetUserRoom).toHaveBeenCalledTimes(1)
})

test('ChatComputer handleNavigation(fn) return link', () => {
  const beverage = {link: "secRoom" }
  const handleNavigation = vi.fn(beverage => {beverage})
  handleNavigation(beverage)
  expect(handleNavigation).toHaveReturned("secRoom")
  expect(handleNavigation).toHaveBeenCalledTimes(1)
})

describe('ChatComputer test of useState', () => {

  let span = HTMLElement

  test("if state change without clicking", () => {
    const setStateMock = vi.fn()
    const useStateMock: any = (useState: any) => [useState, setStateMock]
    vi.spyOn(React, 'useState').mockImplementation(useStateMock)

    const { getByText } = render(
      <MemoryRouter>
        <ChatComputer />
      </MemoryRouter>
    )
    expect(setStateMock).toHaveBeenCalledTimes(0)
  })
})

/*
test('ChatComputer test of useEffect', () => {
  vi.spyOn(React, 'useEffect').mockImplementation((f) => f())
  render(
    <ChatComputer />
  )
  expect().toHaveBeenCalledTimes(1)
})

test("ChatComputer hook testing", () => {
  const mockedNavigate = vi.fn()
  vi.mock('react-router-dom', () => ({
    ...vi.importActual('react-router-dom') as any,
    useNavigate: () => mockedNavigate
  }))

  const mockedParams = vi.fn()
  vi.mock('react-router-dom', () => ({
    ...vi.importActual('react-router-dom') as any,
    useParams: () => ({mockedParams})
  }))

  const mockedAuth = vi.fn()
  vi.mock('react-router-dom', () => ({
    ...vi.importActual('react-router-dom') as any,
    useAuthLogin: () => ({mockedAuth})
  }))
})
*/