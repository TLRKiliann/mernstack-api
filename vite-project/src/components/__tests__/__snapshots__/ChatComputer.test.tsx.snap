// Vitest Snapshot v1

exports[`MatchSnapShot test from ChatComputer 1`] = `
<div
  className="main--chatcomputerroom"
  data-testid="chatcomputer"
>
  <div
    className="div--chatcomputerroom"
    data-testid="chatctestid"
  >
    <img
      alt=""
      className="img--chatcomputerroom"
      height="100%"
      src=""
      width="100%"
    />
  </div>
  <div
    className="title--chatcomputerroom"
  >
    <h1
      className="compter--roomdb"
    />
  </div>
  <div
    className="computer--chatcomputerroom"
  />
</div>
`;
